cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java

```diff
@@ -0,0 +1,183 @@
+    return action;
+  }
+
+  void execute() throws Exception {
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(1, executorTerminateLatch, "action-"
+      + workflowContext.getSpecification().getName() + "-%d");
+
+    try {
+      // Run the action in new thread
+      Future<?> future = executorService.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          setContextCombinedClassLoader(action);
+          try {
+            initializeInTransaction(action, workflowContext);
+            runInTransaction(action, workflowContext);
+            return null;
+          } finally {
+            destroyInTransaction(action, workflowContext);
+          }
+        }
+      });
+      future.get();
+    } catch (Throwable t) {
+      LOG.error("Error executing the action {} in the Workflow {}.", workflowContext.getSpecification().getName(),
+                workflowRunId, t);
+      Throwables.propagateIfPossible(t, Exception.class);
+      throw Throwables.propagate(t);
+    } finally {
+      executorService.shutdownNow();
+      executorTerminateLatch.await();
+    }
+  }
+
+  private void initializeInTransaction(WorkflowAction action, BasicWorkflowContext context) throws Exception {
+    TransactionContext transactionContext = context.getDatasetCache().newTransactionContext();
+    transactionContext.start();
+    try {
+      action.initialize(context);
+    } catch (Throwable t) {
+      LOG.error("Failed to initialize the Workflow action {} for Workflow run {}.",
+                context.getSpecification().getName(), workflowRunId);
+      transactionContext.abort(new TransactionFailureException("Transaction function failure for transaction. ", t));
+    }
+    transactionContext.finish();
+  }
+
+  private void runInTransaction(WorkflowAction action, BasicWorkflowContext context) throws Exception {
+    TransactionContext transactionContext = context.getDatasetCache().newTransactionContext();
+    transactionContext.start();
+    try {
+      action.run();
+    } catch (Throwable t) {
+      LOG.error("Failed to run the Workflow action {} for Workflow run {}.",
+                context.getSpecification().getName(), workflowRunId);
+      transactionContext.abort(new TransactionFailureException("Transaction function failure for transaction. ", t));
+    }
+    transactionContext.finish();
+  }
+
+  private void destroyInTransaction(WorkflowAction action, BasicWorkflowContext context) {
+    TransactionContext transactionContext = context.getDatasetCache().newTransactionContext();
+    try {
+      transactionContext.start();
+      action.destroy();
+      transactionContext.finish();
+    } catch (Throwable t) {
+      LOG.error("Failed to destroy the Workflow action {} for Workflow run {}.",
+                context.getSpecification().getName(), workflowRunId);
+      try {
+        transactionContext.abort();
```

Comment:
The `cause` passed to the `abort` method is only used to rethrow the exception. Since in `destroyTransaction` we should not throw any exception, no need to pass the `t` to abort.